torchok/tasks/simmim.py

```diff
@@ -0,0 +1,141 @@
+        """
+        super().__init__(hparams, inputs)
+
+        # BACKBONE
+        backbones_params = backbone_params or dict()
+        self.backbone = BACKBONES.get(backbone_name)(**backbones_params)
+
+        self.backbone_stride = encoder_stride
+        self.mask_patch_size = mask_patch_size
+        self.mask_ratio = mask_ratio
+
+        # POOLING
+        pooling_params = pooling_params or dict()
+        self.pooling = POOLINGS.get(pooling_name)(in_channels=self.backbone.out_channels, **pooling_params)
+
+        self.decoder = nn.Sequential(
+            nn.Conv2d(
+                in_channels=self.backbone.out_channels,
+                out_channels=self.backbone_stride ** 2 * 3, kernel_size=1),
+            nn.PixelShuffle(self.backbone_stride),
+        )
+
+        self.input_size = self.backbone.img_size[0]
+        self.patch_size = self.backbone.patch_embed.patch_size[0]
+
+        self.rand_size = self.input_size // self.mask_patch_size
+        self.scale = self.mask_patch_size // self.patch_size
+
+        self.token_count = self.rand_size ** 2
+        self.mask_count = int(math.ceil(self.token_count * self.mask_ratio))
+
+        self.hook_handle = self.backbone.patch_embed.register_forward_hook(self.hook)
+        self.backbone.patch_embed.mask_token = nn.Parameter(torch.zeros(1, 1, self.backbone.embed_dim))
+        trunc_normal_(self.backbone.patch_embed.mask_token, mean=0., std=.02)
+
+        mask = torch.zeros(self.rand_size, self.rand_size, dtype=torch.bool, device=self.device, requires_grad=False)
+        mask = mask.repeat_interleave(self.scale, dim=0).repeat_interleave(self.scale, dim=1)
+        self.register_buffer('mask', mask)
+
+    def hook(self, module, inp, out):
+        if self.training:
+            B, L, _ = out.shape
+            mask_tokens = module.mask_token.expand(B, L, -1)
+            w = self.mask.type_as(mask_tokens).flatten()[None, :, None]
+            return out * (1. - w) + mask_tokens * out
+        else:
+            return out
+
+    def forward(self, x: torch.Tensor) -> Tuple[float, torch.Tensor]:
+        """Forward method."""
+        z = self.backbone(x)
+        embeddings = self.pooling(z)
+
+        return embeddings
+
+    def _prepare_mask(self):
+        mask = torch.zeros(self.token_count, dtype=torch.bool, device=self.device, requires_grad=False)
+        if self.training:
+            mask_idx = torch.randperm(self.token_count)[:self.mask_count]
+            mask[mask_idx] = True
+
+        mask = mask.reshape((self.rand_size, self.rand_size))
+        mask = mask.repeat_interleave(self.scale, dim=0).repeat_interleave(self.scale, dim=1)
+        self.register_buffer('mask', mask)
+        return mask
+
+    def forward_with_gt(self, batch: Dict[str, Union[Tensor, int]]) -> Dict[str, Tensor]:
+        """Forward with ground truth labels."""
+        input_data = batch.pop('image')
+
+        mask = self._prepare_mask()
+        z = self.backbone(input_data)
```

Comment:
Don't understand where the mask is applied to input_data. May you explain please?